alacritty_terminal/src/term/mod.rs

```diff
@@ -687,6 +687,87 @@ impl VisualBell {
     }
 }
 
+/// LineDamage describes the current damage of a line, represented as two
+/// column coordinates describing the edges of the damage. Damage is valid if
+/// left <= right.
+#[derive(Clone)]
+pub struct LineDamage {
+    pub left: Column,
+    pub right: Column,
+    pub line: Line,
+}
+
+impl LineDamage {
+    /// Create a new invalid LineDamage, which has left=columns+1 and right=0.
+    /// This ensures that later update with min/max of any valid column
+    /// automatically creates a valid line damage without needing any
+    /// conditionals.
+    fn new(line: Line, columns: Column) -> LineDamage {
+        LineDamage{
+            left: columns+1,
+            right: Column(0),
+            line,
+        }
+    }
+
+    #[inline(always)]
+    pub fn reset(&mut self, columns: Column) {
+        self.left = columns+1;
+        self.right = Column(0);
+    }
+}
+
+struct Damage {
+    line_damage: Vec<LineDamage>,
+    full_damage: bool,
+    columns: Column,
+}
+
+impl Damage {
+    #[inline(always)]
+    fn new(lines: Line, columns: Column) -> Damage {
+        let mut damage = Damage{
+            line_damage: Vec::with_capacity(lines.0),
+            full_damage: false,
+            columns: columns,
+        };
+        for line in 0..lines.0 {
+            damage.line_damage.push(LineDamage::new(Line(line), columns));
+        }
+        damage
+    }
+
+    #[inline(always)]
+    fn expand_line_damage(&mut self, line: Line, left: Column, right: Column) {
+        let l = self.line_damage.get_mut(line.0).unwrap();
+        l.left = min(left, l.left);
+        l.right = max(right, l.right);
+    }
+
+    #[inline(always)]
+    fn damage_everything(&mut self) {
+        self.full_damage = true;
+    }
+
+    #[inline(always)]
+    fn clear_damage(&mut self) {
+        for l in &mut self.line_damage {
+            l.reset(self.columns);
+        }
+        self.full_damage = false;
+    }
+
+    #[inline(always)]
+    fn resize_damage(&mut self, lines: Line, columns: Column) {
```

Comment:
maybe end_line, or taking usize instead of Line.

Reply:
I'm okay either way, but I had a million Line/Column deref's in the previous implementation, and was under the impression that Line/Column was also used for quantity (e.g. `Grid::num_cols`).

Reply:
Ok, let just keep it this way for now.